hbase-server/src/main/java/org/apache/hadoop/hbase/io/HalfStoreFileReader.java

```diff
@@ -349,4 +353,42 @@ public long getFilterEntries() {
     // Estimate the number of entries as half the original file; this may be wildly inaccurate.
     return super.getFilterEntries() / 2;
   }
+
+  /**
+   * Overrides close method to handle cache evictions for the referred file. If evictionOnClose is
+   * true, we will seek to the block containing the splitCell and evict all blocks from offset 0 up
+   * to that block offset if this is a bottom half reader, or the from the split block offset up to
+   * the end of the file if this is a top half reader.
+   * @param evictOnClose true if it should evict the file blocks from the cache.
+   */
+  @Override
+  public void close(boolean evictOnClose) throws IOException {
+    if (closed.compareAndSet(false, true)) {
+      if (evictOnClose && StoreFileInfo.isReference(this.reader.getPath())) {
+        final HFileReaderImpl.HFileScannerImpl s =
```

Comment:
Usually it is not a good idea to the implementation class directly at upper layer. Why we need to use HFileScannerImpl here?

Reply:
We need to seek to the split cell, in order to calculate offsets for the cache keys of referred file to evict.

Reply:
But there is a seekTo method in HFileScanner interface? I just mean why we need the cast here?

Reply:
I need the HFileScannerImpl.getCurBlock for figuring out the offset.